xen/arch/arm/coproc/plat/coproc_xxx/coproc_xxc.c

```diff
@@ -0,0 +1,227 @@
+/*
+ * xen/arch/arm/coproc/plat/coproc_xxc.c
+ *
+ * COPROC_XXC platform specific code
+ * This is an example of description of a complex coprocessor platform code.
+ * This is an example of a coprocessor description which shares common code
+ * with another coprocessor.
+ *
+ * Andrii Anisov <Andrii_Anisov@epam.com>
+ * Copyright (C) 2016 EPAM Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include <xen/init.h>
+#include <xen/err.h>
+#include <xen/irq.h>
+
+#include "coproc_xxx.h"
+
+#define DT_MATCH_COPROC_XXC DT_MATCH_COMPATIBLE("vendor_xxx,coproc_xxc")
+
+static int vcoproc_xxc_op_write(struct vcpu *v, mmio_info_t *info, register_t r,
+                                void *priv);
+static int vcoproc_xxc_mmu_write(struct vcpu *v, mmio_info_t *info,
+                                 register_t r, void *priv);
+
+static struct mmio_handler_ops vcoproc_xxc_op_mmio_handler = {
+    .read = vcoproc_xxx_read,
+    .write = vcoproc_xxc_op_write,
+};
+
+static struct mmio_handler_ops vcoproc_xxc_mmu_mmio_handler = {
+    .read = vcoproc_xxx_read,
+    .write = vcoproc_xxc_mmu_write,
+};
+
+static struct mmio_handler_ops vcoproc_xxc_sram_mmio_handler = {
+    .read = vcoproc_xxx_read,
+    .write = vcoproc_xxx_write,
+};
+
+static struct pcoproc_mmio coproc_xxc_mmio[] = {
+    {
+        .name = "op",
+        .size = 0x1000,
+        .ops = &vcoproc_xxc_op_mmio_handler,
+    },
+    {
+        .name = "mmu",
+        .size = 0x1000,
+        .ops = &vcoproc_xxc_mmu_mmio_handler,
+    },
+    {
+        .name = "sram",
+        .size = 0,
+        .ops = &vcoproc_xxc_sram_mmio_handler,
+    },
+};
+
+enum COPROC_XXC_IRQ
+{
+    COPROC_XXC_OP = 0,
+    COPROC_XXC_MMU
+};
+
+static struct pcoproc_irq coproc_xxc_irq[] = {
+    {
+        .name = "op",
+        .handler = coproc_xxx_irq_handler,
+    },
+    {
+        .name = "mmu",
+        .handler = coproc_xxx_irq_handler,
+    },
+};
+
+static const struct pcoproc_desc coproc_xxc_desc = {
+    .p_mmio_num = sizeof(coproc_xxc_mmio)/sizeof(struct pcoproc_mmio),
+    .p_mmio = coproc_xxc_mmio,
+    .p_irq_num = sizeof(coproc_xxc_irq)/sizeof(struct pcoproc_irq),
+    .p_irq = coproc_xxc_irq,
+};
```

Comment:
I think that some macro should be introduced to simplify sizeof(coproc_xxc_mmio)/sizeof(struct pcoproc_mmio),  sizeof(coproc_xxc_irq)/sizeof(struct pcoproc_irq).